### PR TITLE
[SPIKE] Decouple UI interaction from wallet package

### PR DESCRIPTION
**Pull request by snario**
This is a spike to explore how the wallet implementation can be split up to exclude the current tight coupling from the UI interaction.

Since more logic than less of the wallet is actually what we want from the proposed `engine` package, the package `wallet` is renamed in this PR to `engine` with the intent to extract the non-engine specific functionality out.

More concretely, this means refactoring the current implementation to not be using the `PureComponent` elements from `react`.

The following `PureComponents` are being used right now as the bridge between the `engine` and the UI. **Since this is not a small change, to prevent major divergence from master, the path of action is to keep the build and tests passing at all times while incrementally refactoring each of the following components separately** (note that the order of presentation isn't necessarily the order in which the items will be checked off; the order will be decided by whatever is easiest to refactor moving forward as this is a spike). It will likely be the case that each component will be tackled in a separate, stacked PR to make reviews easier.

The work for each protocol will be marked as either "IN PROGRESS" or "DONE".

- [ ] defunding
- [ ] concluding
- [ ] funding
- [ ] dispute/challenger
- [ ] dispute/challenger/components/wait-for-response-or-timeout
- [ ] dispute/responder
- [ ] dispute/responder/components/wait-for-approval
- [ ] funding-strategy-negotiation/player-a
- [ ] funding-strategy-negotiation/player-b
- [ ] ledger-funding [IN PROGRESS]
- [ ] ledger-defunding
- [ ] virtual-defunding
- [ ] close-ledger-channel
- [ ] withdrawing
- [ ] new-ledger-channel
- [ ] existing-ledger-funding
- [ ] consensus-update
- [ ] advance-channel
- [ ] withdrawing/components/wait-for-approval
- [ ] ledger-top-up
- [ ] new-ledger-channel/components/funding-step
- [ ] advance-channel/components/advancing-channel
- [ ] direct-funding
- [ ] direct-funding/components/funding-step
- [ ] src/containers/engine

### CODE REVIEW

**Comment by lalexgap** [comment]
> Do you guys think there is a way to make this change as a series of PRs into master as opposed to one giant PR? Perhaps we can:
> 
> 1. Change `wallet` to `engine` naming scheme.
> 2. Iteratively split engine logic from UI for each protocol.

I would be in favor of that approach!

File: packages/engine/contracts/pre-built-artifacts/Migrations.json [hunk 1]
```diff
@@ -68,15 +68,15 @@
       "type": "function"
     }
   ],
```
[the remark this answers is not in the document]
Comment (lalexgap) [reply packages/engine/contracts/pre-built-artifacts/Migrations.json]: > Perhaps this is not immediately relevant to this work, but I am trying to remember why pre-built-artifacts are part of the `wallet` package. Should these artifacts live in the `nitro-protocol` package?

I think we should remove them for now. If we want to get rid of them we'll have to remove this webpack plugin:
```
  new webpack.NormalModuleReplacementPlugin(
        /.*\/build\/contracts\/.*\.json/,
        function (resource) {
          if (process.env.TARGET_NETWORK !== 'development') {
            resource.request = resource.request.replace(/.*\/build\/contracts/, paths.appPreBuiltContractArtifacts);
          }
        }),```

**Comment by cf19drofxots** [comment]
> Do you guys think there is a way to make this change as a series of PRs into master as opposed to one giant PR? 

I am in favor of this. 

> Iteratively split engine logic from UI for each protocol.

This would be the same as having the stacked PRs against this branch, except against master (which would also be better to reduce merge conflicts).

**Comment by cf19drofxots** [comment]
Ok looks like @lalexgap also agrees. Will just merge this as is then and move the list of check boxes to an issue to track this work as an epic.

File: packages/engine/contracts/pre-built-artifacts/Migrations.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (cf19drofxots) [reply packages/engine/contracts/pre-built-artifacts/Migrations.json]: Is this something that should be done in this PR? (My guess is no, but asking for confirmation)

[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/engine/contracts/pre-built-artifacts/Migrations.json]: I filed https://github.com/statechannels/monorepo/issues/319. We can ignore that for now.

**Pull request merged**